BV2-169 (Feat) : SSE 개발

seob7 opened the pull request:
## #️⃣ Issue Number

N/A

## 📝 요약(Summary)

- `SseEmitterManager` 개발
    -  SSE Connection time out = 1h
    -  생성된 Emitter 는 Local Memory 저장
    - 메시지 수신 테스트를 위해 `SseMessageManager` 도 개발했습니다.
- 입/퇴장 시 redis 에서 userId 추가/제거하도록 함
- 메시지 전송 시 Local Memory 에 SSE 정보가 존재하지 않다면, Redis 에서 접속한 사용자의 서버 주소를 파악 후 메시지 전송
    - key : value = userID : Server Instance Address
- `NormalTitleGenerator`, `TruncatedTitleGenerator` 싱글톤 패턴 적용

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 공유사항 to 리뷰어

<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## 📑 API Spec

| **기능**          |                |
|-----------------|----------------|
| **HTTP Method** |                |
| **URL**         |                |

### Request Body

### Response Body


## 📸스크린샷 (선택)


## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다. (버그 수정/기능에 대한 테스트)

seob7 committed:
BV2-169 (Feat) : SSE 개발

BV2-169 (Feat) : SSE 개발

1. SseEmitterManager
2. SseMessageManager

seob7 committed:
BV2-169 (Refactor) : Adapter-Repository 구분

BV2-169 (Refactor) : Adapter-Repository 구분

seob7 committed:
BV2-169 (Feat) : 방생성/퇴장 기능 개발

BV2-169 (Feat) : 방생성/퇴장 기능 개발

seob7 committed:
BV2-169 (Refactor) : connect -> create 메서드명 변경

BV2-169 (Refactor) : connect -> create 메서드명 변경

1. connect -> create 으로 변경

shinywoon commented on chat/src/main/java/chat/adapter/output/SseConnectionPoolAdapter.java:
```diff
@@ -0,0 +1,29 @@
+@PersistenceAdapter
+public class SseConnectionPoolAdapter implements SseConnectionPoolPort<String, UserSseConnection> {
+
+    private static final Map<String, UserSseConnection> connectionPool = new ConcurrentHashMap<>();
```
🧐SseConnectionPool 이라고 자료구조 명칭을 지어주셨는데...
Connection을 반환하거나 공유하지 않으니 Flyweight Pattern 의 Pool 개념과 충돌하는 것 같아요!

shinywoon commented on chat/src/main/java/chat/adapter/output/SseConnectionPoolAdapter.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void addSession(String userId, UserSseConnection connection) {
+        connectionPool.put(userId, connection);
+    }
```
Session 을 `더한다` 는 개념보다는 Connection 을 저장한다는 의미가 더 맞는 표현인것 같아요!

shinywoon commented on chat/src/main/java/chat/adapter/output/SseConnectionPoolAdapter.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void removeSession(UserSseConnection session) {
+        connectionPool.remove(session.getUserId());
+    }
```
`getEmitter` `removeEmitter` 겟죠?
Session 을 사용하고 싶다면! User Session 을 저장하는 정보에 사용하는게 좋을것 같아요!

shinywoon commented on chat/src/main/java/chat/application/sse/SseEmitterManager.java:
```diff
@@ -0,0 +1,20 @@
+        UserSseConnection userSseConnection = UserSseConnection.connect(userId, connectionPoolPort);
+        connectionPoolPort.addSession(userId, userSseConnection);
+       return userSseConnection;
+    }
```
저희 채팅의 Emitter 활용의 경우를 보면
<Server>
Chat Room 입장 요청 접수 >  Emitter 생성 > Emitter 반환
Message Event 발행
<Client>
Chat Room 입장 요청 > Emitter 수신 > Emitter 연결 
Message Event 수신
입니다. 
즉 Connection을 하는것은 저희의 역할이 아니며 그저 Emitter 생성 및 반환이 전부 입니다.
Logic 의 수행 결과는 위와 같으나 
Logic 의 명칭은 Client 의 역할을 수행하는 것으로 보입니다.

shinywoon commented on chat/src/main/java/chat/application/sse/UserSseConnection.java:
```diff
@@ -0,0 +1,41 @@
+        SseConnectionPoolPort<String, UserSseConnection> connectionPoolPort
+    ){
+        return new UserSseConnection(userId, connectionPoolPort);
+    }
```
CallBack 활용과 코드의 구조는 정말 좋네요!

shinywoon reviewed (commented):
전체적인 코드의 구성은 좋은것 같습니다.
다만 Method명이나 변수명이 역할에서 많이 벗어나 있다는 점이 아쉽네요.
하는 일을 명확히 설명할 수 있도록 노력해주세요!

seob7 committed:
BV2-169 (Refactor) : 클래스, 메서드명 변경

BV2-169 (Refactor) : 클래스, 메서드명 변경

1. SseConnectionPoolXXXX -> SseConnectionRegistryXXXX
2. xxxSession -> xxxEmitter 로 변경

seob7 committed:
BV2-169 (Refactor) : `createSseEmitter` emitter 생성만 하도록 변경

BV2-169 (Refactor) : `createSseEmitter` emitter 생성만 하도록 변경

1. createSseEmitter 는 Emiitter 생성만 허도록 변경
2. connectionPoolPort -> onnectionRegistryPort 변수명 변경

seob7 committed:
BV2-169 (Refactor) : 채팅방 입장시 emitter 생성, 사용자 정보 저장하도록 변경

BV2-169 (Refactor) : 채팅방 입장시 emitter 생성, 사용자 정보 저장하도록 변경

seob7 committed:
BV2-169 (Refactor) : Redis 에 userId:server address 저장하도록 변경

BV2-169 (Refactor) : Redis 에 userId:server address 저장하도록 변경

- 채팅방 접속 시 Redis 에 `userId : server address` 저장

seob7 committed:
BV2-259 (Refactor) : XXXTitleGenerator 싱글톤으로 변경

BV2-259 (Refactor) : XXXTitleGenerator 싱글톤으로 변경
